resource_work_time_from_contracts/models/resource_mixin.py

```diff
@@ -104,56 +35,209 @@ def get_work_days_data(
         calendar=None,
         domain=None,
     ):
-        if calendar or not hasattr(self, "contract_ids"):
+        # sudo() is needed for normal users that have no read access to
+        # contracts because hasattr() tries to access the field's value.
+        if calendar or not hasattr(self.sudo(), "contract_ids"):
             return super().get_work_days_data(
                 from_datetime, to_datetime, compute_leaves, calendar, domain
             )
         # we need the normal work time per day for each day to be able to
-        # compute the fraction of day that the number of hours represents.
-        # this is defined in the calendar of the employee, which must be the
-        # same as the one of the company, and which must contain full-day
-        # hours for each day.
+        # compute the fraction of day that the number of hours represents. we
+        # choose to use the resource.calendar.hours_per_day field for this. it
+        # is automatically computed, but can be edited. this way, it is
+        # possible to have irregular work schedules, for example where some
+        # days only have a half day, and taking a leave on these will be
+        # counted as a half day instead of as a full day.
         #
         # we need full days, so we replace the hours to start and stop at
         # midnight in the timezone of the resource.
-        #
-        # the provided domain is to exclude leaves from the computation, as
-        # explained in list_work_time_per_day().
         from_datetime, to_datetime = self._localize_datetimes(
             from_datetime, to_datetime
         )
+        normal_attendance_intervals = self._get_attendance_intervals(
+            from_datetime.replace(hour=0, minute=0, second=0, microsecond=0),
+            to_datetime.replace(hour=0, minute=0, second=0, microsecond=0)
+            + datetime.timedelta(days=1),
+        )
         normal_work_time_per_day = dict(
-            super().list_work_time_per_day(
-                from_datetime.replace(hour=0, minute=0, second=0, microsecond=0),
-                to_datetime.replace(hour=0, minute=0, second=0, microsecond=0)
-                + datetime.timedelta(days=1),
-                self.resource_calendar_id,
-                [("calendar_id", "=", False)],
-            )
+            self._hours_per_day_from_intervals(normal_attendance_intervals)
         )
         if compute_leaves:
             work_time_per_day = self.list_work_time_per_day(
                 from_datetime, to_datetime, calendar=None, domain=domain
             )
         else:
             work_time_per_day = self.list_normal_work_time_per_day(
-                from_datetime, to_datetime, domain=domain
+                from_datetime, to_datetime
+            )
+        return self._sum_work_days_data(work_time_per_day, normal_work_time_per_day)
+
```

Comment:
The method called here is unique to this method, and should be closer.

Reply:
moved it as close as possible (while keeping private methods after public ones, and sub-methods above the ones that use them).